src/main/scala/components/memory-async.scala

```diff
@@ -0,0 +1,180 @@
+// Asynchronous memory module
+
+package dinocpu
+
+import chisel3._
+import chisel3.util._
+
+import chisel3.util.experimental.loadMemoryFromFile
+import firrtl.annotations.MemoryLoadFileType
+
+//
+/**
+ * The *interface* for the asynchronous instruction memory. All inputs of the synchronous IMemIO bundle
+ * are copied into this bundle, along with two extra flags:  
+ *
+ * Input: ready, true when an address is being supplied to the instruction memory, indicating that an instruction should be fetched
+ *
+ * Output: valid, true when the memory is outputting a valid instruction through imem.instruction 
+ */
+class AsyncIMemIO extends IMemIO {
+  val ready = Input(Bool())
+
+  val valid = Output(Bool())
+}
+
+/**
+ * The *interface* for the asynchronous data memory. All inputs of the synchronous DMemIO bundle
+ * are copied into this bundle along with an extra flag:
+ * 
+ * Output: valid, true when the number of cycles that the async memory is supposed to wait for 
+ *         elapses, and readdata contains the data that was read and processed 
+ *
+ * The 'readiness' of the data address input is inferred from whether dmem.memread or dmem.memwrite
+ * is true, as otherwise ready would be true when dmem.memread or dmem.memwrite is true.
+ */
+class AsyncDMemIO extends DMemIO {
+  val valid = Output(Bool())
+}
+
+// A Bundle used for storing inputs while memory is being delayed.
+class AsyncStoredIO extends Bundle {  
+  val address   = UInt(32.W)
+  val writedata = UInt(32.W)
+  val maskmode  = UInt(2.W)
+  val sext      = Bool()
+  val delaying  = Bool()
+
+  // operation is an encoding of the current memory operation, represented below:
+  //             00 -> Instruction read
+  //             01 -> Data read
+  //             10 (-> Data write
+  // Maybe use an enumerator for this, if possible?
+  val operation = UInt(2.W)
```

Comment:
Why do we need a different operation for instruction and data reading?

Reply:
I needed some kind of way to indicate that after accessing memory that either the instruction port's or the data port's output is valid. If I just encoded the read operation by itself then I believe there would be no way for the memory to distinguish whether it should be sending an instruction or a piece of data, so we would have to send the data over to both ports and say they're both valid.

But I'm thinking that that won't be a problem after all since the pipeline will only either read instructions or data, but never both at the same time. We would need to make a clarification though that the output of the data memory and instruction memory are in fact the same and nobody should expect them to have different values for any reason.